src/transformStateWithClones.js

```diff
@@ -6,8 +6,35 @@
  *
  * @return {Object[]}
  */
+
 function transformStateWithClones(state, actions) {
-  // write code here
+  let stateClone = Object.assign({}, state);
+  const allStates = [];
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        Object.assign(stateClone, action.extraData);
+        break;
+
+      case 'removeProperties':
+        for (const eachKey of action.keysToRemove) {
+          delete stateClone[eachKey];
+        }
+        break;
+
+      case 'clear':
+        stateClone = {};
+        break;
+
+      default:
+        break;
+    }
+
+    allStates.push(Object.assign({}, stateClone));
```

Comment:
good but can be more simplify
```suggestion
    allStates.push({ ...stateClone });
```